File: web-app/src/api/analysis.ts
import { analysisApi } from "./sdk";
import { GetIncomeAnalysisApiAnalysisIncomeGetMode } from "@spearmint-finance/sdk";

// Analysis API functions

export interface IncomeAnalysisResponse {
  total_income: number;
  transaction_count: number;
  average_transaction: number;
  breakdown_by_category: Record<string, CategoryBreakdown>;
  period_start: string | null;
  period_end: string | null;
  mode: "analysis" | "complete";
}

export interface ExpenseAnalysisResponse {
  total_expenses: number;
  transaction_count: number;
  average_transaction: number;
  breakdown_by_category: Record<string, CategoryBreakdown>;
  top_categories: TopCategory[];
  period_start: string | null;
  period_end: string | null;
  mode: "analysis" | "complete";
}

export interface CategoryBreakdown {
  total: number;
  count: number;
  average: number;
  percentage: number;
}

export interface TopCategory {
  category: string;
  amount: number;
  count: number;
  percentage: number;
}

export interface CashFlowResponse {
  total_income: number;
  total_expenses: number;
  net_cash_flow: number;
  income_count: number;
  expense_count: number;
  period_start: string | null;
  period_end: string | null;
  mode: "analysis" | "complete";
}

export interface FinancialHealthResponse {
  income_to_expense_ratio: number | null;
  savings_rate: number | null;
  average_daily_income: number;
  average_daily_expense: number;
  net_daily_cash_flow: number;
  period_start: string | null;
  period_end: string | null;
}

export interface FinancialSummaryResponse {
  total_income: number;
  total_expenses: number;
  net_cash_flow: number;
  income_count: number;
  expense_count: number;
  top_income_categories: TopCategory[];
  top_expense_categories: TopCategory[];
  recent_transactions: RecentTransaction[];
  financial_health: FinancialHealthResponse;
  period_start: string | null;
  period_end: string | null;
  mode: "analysis" | "complete";
}

export interface RecentTransaction {
  transaction_id: number;
  transaction_date: string;
  description: string;
  amount: number;
  transaction_type: "Income" | "Expense";
  category: string | null;
}

export interface AnalysisParams {
  start_date?: string;
  end_date?: string;
  mode?: "analysis" | "with_capital" | "complete";
}

export interface CashFlowTrendPoint {
  period: string;
  income: number;
  expenses: number;
  net_cash_flow: number;
  income_count: number;
  expense_count: number;
}

export interface CashFlowTrendsResponse {
  trends: CashFlowTrendPoint[];
  period_type: "daily" | "weekly" | "monthly" | "quarterly" | "yearly";
  mode: "analysis" | "complete";
}

// Helper to convert string mode to Enum
const toModeEnum = (
  mode?: string
): GetIncomeAnalysisApiAnalysisIncomeGetMode | undefined => {
  if (mode === "analysis")
    return GetIncomeAnalysisApiAnalysisIncomeGetMode.ANALYSIS;
  if (mode === "complete")
    return GetIncomeAnalysisApiAnalysisIncomeGetMode.COMPLETE;
  if (mode === "with_capital")
    return GetIncomeAnalysisApiAnalysisIncomeGetMode.WITH_CAPITAL;
  return undefined;
};

/**
 * Get income analysis
 */
export const getIncomeAnalysis = async (
  params?: AnalysisParams
): Promise<IncomeAnalysisResponse> => {
  const { data } = await analysisApi.getIncomeAnalysisApiAnalysisIncomeGet({
    startDate: params?.start_date,
    endDate: params?.end_date,
    mode: toModeEnum(params?.mode),
  });
  if (!data) {
    throw new Error("Empty response from income analysis API");
  }
  return {
    total_income: Number(data.totalIncome),
    transaction_count: data.transactionCount,
    average_transaction: Number(data.averageTransaction),
    breakdown_by_category: Object.fromEntries(
      Object.entries(data.breakdownByCategory ?? {}).map(
        ([key, value]: any) => [
          key,
          {
            total: Number(value.total),
            count: value.count,
            average: Number(value.average),
            percentage: value.percentage,
          },
        ]
      )
    ),
    period_start: data.periodStart ?? null,
    period_end: data.periodEnd ?? null,
    mode: (data.mode as "analysis" | "complete") ?? "analysis",
  };
};

/**
 * Get expense analysis
 */
export const getExpenseAnalysis = async (
  params?: AnalysisParams & { top_n?: number }
): Promise<ExpenseAnalysisResponse> => {
  const { data } = await analysisApi.getExpenseAnalysisApiAnalysisExpensesGet({
    startDate: params?.start_date,
    endDate: params?.end_date,
    mode: toModeEnum(params?.mode),
    topN: params?.top_n,
  });
  if (!data) {
    throw new Error("Empty response from expense analysis API");
  }
  return {
    total_expenses: Number(data.totalExpenses),
    transaction_count: data.transactionCount,
    average_transaction: Number(data.averageTransaction),
    breakdown_by_category: Object.fromEntries(
      Object.entries(data.breakdownByCategory ?? {}).map(
        ([key, value]: any) => [
          key,
          {
            total: Number(value.total),
            count: value.count,
            average: Number(value.average),
            percentage: value.percentage,
          },
        ]
      )
    ),
    top_categories: (data.topCategories ?? []).map((cat) => ({
      category: cat.category,
      amount: Number(cat.amount),
      count: cat.count,
      percentage: cat.percentage,
    })),
    period_start: data.periodStart ?? null,
    period_end: data.periodEnd ?? null,
    mode: (data.mode as "analysis" | "complete") ?? "analysis",
  };
};

/**
 * Get cash flow analysis
 */
export const getCashFlowAnalysis = async (
  params?: AnalysisParams
): Promise<CashFlowResponse> => {
  const { data } = await analysisApi.getCashFlowAnalysisApiAnalysisCashflowGet({
    startDate: params?.start_date,
    endDate: params?.end_date,
    mode: toModeEnum(params?.mode),
  });
  if (!data) {
    throw new Error("Empty response from cash flow analysis API");
  }
  return {
    total_income: Number(data.totalIncome),
    total_expenses: Number(data.totalExpenses),
    net_cash_flow: Number(data.netCashFlow),
    income_count: data.incomeCount,
    expense_count: data.expenseCount,
    period_start: data.periodStart ?? null,
    period_end: data.periodEnd ?? null,
    mode: (data.mode as "analysis" | "complete") ?? "analysis",
  };
};

/**
 * Get financial health indicators
 */
export const getFinancialHealth = async (
  params?: AnalysisParams
): Promise<FinancialHealthResponse> => {
  const { data } = await analysisApi.getFinancialHealthApiAnalysisHealthGet({
    startDate: params?.start_date,
    endDate: params?.end_date,
    mode: toModeEnum(params?.mode),
  });
  if (!data) {
    throw new Error("Empty response from financial health API");
  }
  return {
    income_to_expense_ratio: data.incomeToExpenseRatio ?? null,
    savings_rate: data.savingsRate ?? null,
    average_daily_income: Number(data.averageDailyIncome),
    average_daily_expense: Number(data.averageDailyExpense),
    net_daily_cash_flow: Number(data.netDailyCashFlow),
    period_start: data.periodStart ?? null,
    period_end: data.periodEnd ?? null,
  };
};

/**
 * Get financial summary (comprehensive dashboard data)
 */
export const getFinancialSummary = async (
  params?: AnalysisParams & { top_n?: number; recent_count?: number }
): Promise<FinancialSummaryResponse> => {
  const { data } = await analysisApi.getFinancialSummaryApiAnalysisSummaryGet({
    startDate: params?.start_date,
    endDate: params?.end_date,
    mode: toModeEnum(params?.mode),
    topN: params?.top_n,
    recentCount: params?.recent_count,
  });
  if (!data) {
    throw new Error("Empty response from financial summary API");
  }
  return {
    total_income: Number(data.totalIncome),
    total_expenses: Number(data.totalExpenses),
    net_cash_flow: Number(data.netCashFlow),
    income_count: data.incomeCount,
    expense_count: data.expenseCount,
    top_income_categories: (data.topIncomeCategories ?? []).map((cat) => ({
      category: cat.category,
      amount: Number(cat.amount),
      count: cat.count,
      percentage: cat.percentage,
    })),
    top_expense_categories: (data.topExpenseCategories ?? []).map((cat) => ({
      category: cat.category,
      amount: Number(cat.amount),
      count: cat.count,
      percentage: cat.percentage,
    })),
    recent_transactions: (data.recentTransactions ?? []).map((txn) => ({
      transaction_id: txn.transactionId,
      transaction_date: txn.transactionDate,
      description: txn.description,
      amount: Number(txn.amount),
      transaction_type: txn.transactionType as "Income" | "Expense",
      category: txn.category ?? null,
    })),
    financial_health: {
      income_to_expense_ratio:
        data.financialHealth.incomeToExpenseRatio ?? null,
      savings_rate: data.financialHealth.savingsRate ?? null,
      average_daily_income: Number(data.financialHealth.averageDailyIncome),
      average_daily_expense: Number(data.financialHealth.averageDailyExpense),
      net_daily_cash_flow: Number(data.financialHealth.netDailyCashFlow),
      period_start: data.financialHealth.periodStart ?? null,
      period_end: data.financialHealth.periodEnd ?? null,
    },
    period_start: data.periodStart ?? null,
    period_end: data.periodEnd ?? null,
    mode: (data.mode as "analysis" | "complete") ?? "analysis",
  };
};

export const getCashFlowTrends = async (
  params?: AnalysisParams & {
    period?: "daily" | "weekly" | "monthly" | "quarterly" | "yearly";
  }
): Promise<CashFlowTrendsResponse> => {
  const { data } =
    await analysisApi.getCashFlowTrendsApiAnalysisCashflowTrendsGet({
      startDate: params?.start_date,
      endDate: params?.end_date,
      mode: toModeEnum(params?.mode),
      period: params?.period as any, // Cast to TimePeriodEnum if available
    });
  if (!data) {
    throw new Error("Empty response from cash flow trends API");
  }
  return {
    trends: (data.trends ?? []).map((t) => ({
      period: t.period,
      income: Number(t.income),
      expenses: Number(t.expenses),
      net_cash_flow: Number(t.netCashFlow),
      income_count: t.incomeCount,
      expense_count: t.expenseCount,
    })),
    period_type: data.periodType as any,
    mode: (data.mode as "analysis" | "complete") ?? "analysis",
  };
};

/**
 * Get expense category trends for stacked bar charts
 */
export interface CategoryTrendsParams {
  start_date?: string;
  end_date?: string;
  period?: "daily" | "weekly" | "monthly" | "quarterly" | "yearly";
  mode?: "analysis" | "with_capital" | "complete";
  top_n?: number;
}

export interface CategoryTrendsResponse {
  periods: string[];
  categories: string[];
  data: Array<{
    period: string;
    [category: string]: string | number;
  }>;
  period_type: string;
  mode: string;
}

export const getExpenseCategoryTrends = async (
  params: CategoryTrendsParams
): Promise<CategoryTrendsResponse> => {
  const { data } =
    await analysisApi.getExpenseCategoryTrendsApiAnalysisExpensesCategoryTrendsGet(
      {
        startDate: params.start_date,
        endDate: params.end_date,
        period: params.period as any,
        mode: toModeEnum(params.mode),
        topN: params.top_n,
      }
    );
  if (!data) {
    throw new Error("Empty response from expense category trends API");
  }
  return data as CategoryTrendsResponse;
};

File: core-api/src/financial_analysis/api/routes/analysis.py
"""Analysis API endpoints.

Provides endpoints for income, expense, cash flow, and financial health analysis.  

"""

from typing import Optional
from datetime import date
from fastapi import APIRouter, Depends, Query
from sqlalchemy.orm import Session

from ..dependencies import get_db
from ..schemas.analysis import (
    DateRangeRequest,
    IncomeAnalysisResponse,
    ExpenseAnalysisResponse,
    CashFlowResponse,
    TrendsResponse,
    CashFlowTrendsResponse,
    FinancialHealthResponse,
    PeriodComparisonResponse,
    TimePeriodEnum,
    AnalysisModeEnum,
    CategoryBreakdown,
    TopCategory,
    TrendDataPoint,
    CashFlowTrendPoint,
    FinancialSummaryResponse,
    RecentTransaction,
    IncomeExpenseComparisonResponse,
    CategoryBreakdownResponse,
    CategoryBreakdownItem
)
from ...services.analysis_service import (
    AnalysisService,
    DateRange,
    AnalysisMode,
    TimePeriod
)

router = APIRouter()


def _convert_mode(mode: AnalysisModeEnum) -> AnalysisMode:
    """Convert API mode enum to service mode enum."""
    if mode == AnalysisModeEnum.ANALYSIS:
        return AnalysisMode.ANALYSIS
    elif mode == AnalysisModeEnum.WITH_CAPITAL:
        return AnalysisMode.WITH_CAPITAL
    else:
        return AnalysisMode.COMPLETE


def _convert_period(period: TimePeriodEnum) -> TimePeriod:
    """Convert API period enum to service period enum."""
    return TimePeriod(period.value)


@router.get("/analysis/income", response_model=IncomeAnalysisResponse)
def get_income_analysis(
    start_date: Optional[date] = Query(None, description="Start date for analysis"),
    end_date: Optional[date] = Query(None, description="End date for analysis"),
    mode: AnalysisModeEnum = Query(AnalysisModeEnum.ANALYSIS, description="Analysis mode"),
    entity_id: Optional[int] = Query(None, description="Entity ID to filter by"),
    db: Session = Depends(get_db)
):
    """
    Get income analysis for a period.

    Args:
        start_date: Start date for analysis
        end_date: End date for analysis
        mode: Analysis mode (analysis or complete)
        entity_id: Entity ID to filter by
        db: Database session

    Returns:
        IncomeAnalysisResponse: Income analysis results
    """
    service = AnalysisService(db)

    date_range = DateRange(start_date=start_date, end_date=end_date) if start_date or end_date else None
    result = service.analyze_income(date_range=date_range, mode=_convert_mode(mode), entity_id=entity_id)
    
    # Convert breakdown to Pydantic models
    breakdown = {
        cat: CategoryBreakdown(**data)
        for cat, data in result.breakdown_by_category.items()
    }
    
    return IncomeAnalysisResponse(
        total_income=result.total_income,
        transaction_count=result.transaction_count,
        average_transaction=result.average_transaction,
        breakdown_by_category=breakdown,
        period_start=result.period_start,
        period_end=result.period_end,
        mode=AnalysisModeEnum(result.mode.value)
    )


@router.get("/analysis/income/trends", response_model=TrendsResponse)
def get_income_trends(
    start_date: Optional[date] = Query(None, description="Start date for analysis"),
    end_date: Optional[date] = Query(None, description="End date for analysis"),
    period: TimePeriodEnum = Query(TimePeriodEnum.MONTHLY, description="Period granularity"),
    mode: AnalysisModeEnum = Query(AnalysisModeEnum.ANALYSIS, description="Analysis mode"),
    entity_id: Optional[int] = Query(None, description="Entity ID to filter by"),
    db: Session = Depends(get_db)
):
    """
    Get income trends over time.

    Args:
        start_date: Start date for analysis
        end_date: End date for analysis
        period: Period granularity (daily, weekly, monthly, etc.)
        mode: Analysis mode
        entity_id: Entity ID to filter by
        db: Database session

    Returns:
        TrendsResponse: Income trend data
    """
    service = AnalysisService(db)

    date_range = DateRange(start_date=start_date, end_date=end_date) if start_date or end_date else None
    trends = service.get_income_trends(
        date_range=date_range,
        period=_convert_period(period),
        mode=_convert_mode(mode),
        entity_id=entity_id
    )
    
    return TrendsResponse(
        trends=[TrendDataPoint(period=t.period, value=t.value, count=t.count) for t in trends],
        period_type=period,
        mode=mode
    )


@router.get("/analysis/expenses", response_model=ExpenseAnalysisResponse)
def get_expense_analysis(
    start_date: Optional[date] = Query(None, description="Start date for analysis"),
    end_date: Optional[date] = Query(None, description="End date for analysis"),
    mode: AnalysisModeEnum = Query(AnalysisModeEnum.ANALYSIS, description="Analysis mode"),
    top_n: int = Query(10, ge=1, le=50, description="Number of top categories to return"),
    entity_id: Optional[int] = Query(None, description="Entity ID to filter by"),
    db: Session = Depends(get_db)
):
    """
    Get expense analysis for a period.

    Args:
        start_date: Start date for analysis
        end_date: End date for analysis
        mode: Analysis mode (analysis or complete)
        top_n: Number of top categories to return
        entity_id: Entity ID to filter by
        db: Database session

    Returns:
        ExpenseAnalysisResponse: Expense analysis results
    """
    service = AnalysisService(db)

    date_range = DateRange(start_date=start_date, end_date=end_date) if start_date or end_date else None
    result = service.analyze_expenses(date_range=date_range, mode=_convert_mode(mode), top_n=top_n, entity_id=entity_id)
    
    # Convert breakdown to Pydantic models
    breakdown = {
        cat: CategoryBreakdown(**data)
        for cat, data in result.breakdown_by_category.items()
    }
    
    # Convert top categories
    top_categories = [TopCategory(**cat) for cat in result.top_categories]
    
    return ExpenseAnalysisResponse(
        total_expenses=result.total_expenses,
        transaction_count=result.transaction_count,
        average_transaction=result.average_transaction,
        breakdown_by_category=breakdown,
        top_categories=top_categories,
        period_start=result.period_start,
        period_end=result.period_end,
        mode=AnalysisModeEnum(result.mode.value)
    )


@router.get("/analysis/expenses/trends", response_model=TrendsResponse)
def get_expense_trends(
    start_date: Optional[date] = Query(None, description="Start date for analysis"),
    end_date: Optional[date] = Query(None, description="End date for analysis"),
    period: TimePeriodEnum = Query(TimePeriodEnum.MONTHLY, description="Period granularity"),
    mode: AnalysisModeEnum = Query(AnalysisModeEnum.ANALYSIS, description="Analysis mode"),
    entity_id: Optional[int] = Query(None, description="Entity ID to filter by"),
    db: Session = Depends(get_db)
):
    """
    Get expense trends over time.

    Args:
        start_date: Start date for analysis
        end_date: End date for analysis
        period: Period granularity (daily, weekly, monthly, etc.)
        mode: Analysis mode
        entity_id: Entity ID to filter by
        db: Database session

    Returns:
        TrendsResponse: Expense trend data
    """
    service = AnalysisService(db)

    date_range = DateRange(start_date=start_date, end_date=end_date) if start_date or end_date else None
    trends = service.get_expense_trends(
        date_range=date_range,
        period=_convert_period(period),
        mode=_convert_mode(mode),
        entity_id=entity_id
    )
    
    return TrendsResponse(
        trends=[TrendDataPoint(period=t.period, value=t.value, count=t.count) for t in trends],
        period_type=period,
        mode=mode
    )


@router.get("/analysis/cashflow", response_model=CashFlowResponse)
def get_cash_flow_analysis(
    start_date: Optional[date] = Query(None, description="Start date for analysis"),
    end_date: Optional[date] = Query(None, description="End date for analysis"),
    mode: AnalysisModeEnum = Query(AnalysisModeEnum.ANALYSIS, description="Analysis mode"),
    entity_id: Optional[int] = Query(None, description="Entity ID to filter by"),
    db: Session = Depends(get_db)
):
    """
    Get cash flow analysis for a period.

    Args:
        start_date: Start date for analysis
        end_date: End date for analysis
        mode: Analysis mode (analysis or complete)
        entity_id: Entity ID to filter by
        db: Database session

    Returns:
        CashFlowResponse: Cash flow analysis results
    """
    service = AnalysisService(db)

    date_range = DateRange(start_date=start_date, end_date=end_date) if start_date or end_date else None
    result = service.analyze_cash_flow(date_range=date_range, mode=_convert_mode(mode), entity_id=entity_id)
    
    return CashFlowResponse(
        net_cash_flow=result.net_cash_flow,
        total_income=result.total_income,
        total_expenses=result.total_expenses,
        income_count=result.income_count,
        expense_count=result.expense_count,
        period_start=result.period_start,
        period_end=result.period_end,
        mode=AnalysisModeEnum(result.mode.value)
    )


@router.get("/analysis/cashflow/trends", response_model=CashFlowTrendsResponse)
def get_cash_flow_trends(
    start_date: Optional[date] = Query(None, description="Start date for analysis"),
    end_date: Optional[date] = Query(None, description="End date for analysis"),
    period: TimePeriodEnum = Query(TimePeriodEnum.MONTHLY, description="Period granularity"),
    mode: AnalysisModeEnum = Query(AnalysisModeEnum.ANALYSIS, description="Analysis mode"),
    entity_id: Optional[int] = Query(None, description="Entity ID to filter by"),
    db: Session = Depends(get_db)
):
    """
    Get cash flow trends over time.

    Args:
        start_date: Start date for analysis
        end_date: End date for analysis
        period: Period granularity (daily, weekly, monthly, etc.)
        mode: Analysis mode
        entity_id: Entity ID to filter by
        db: Database session

    Returns:
        CashFlowTrendsResponse: Cash flow trend data
    """
    service = AnalysisService(db)

    date_range = DateRange(start_date=start_date, end_date=end_date) if start_date or end_date else None
    trends = service.get_cash_flow_trends(
        date_range=date_range,
        period=_convert_period(period),
        mode=_convert_mode(mode),
        entity_id=entity_id
    )
    
    return CashFlowTrendsResponse(
        trends=[CashFlowTrendPoint(**t) for t in trends],
        period_type=period,
        mode=mode
    )


@router.get("/analysis/health", response_model=FinancialHealthResponse)
def get_financial_health(
    start_date: Optional[date] = Query(None, description="Start date for analysis"),
    end_date: Optional[date] = Query(None, description="End date for analysis"),
    mode: AnalysisModeEnum = Query(AnalysisModeEnum.ANALYSIS, description="Analysis mode"),
    entity_id: Optional[int] = Query(None, description="Entity ID to filter by"),
    db: Session = Depends(get_db)
):
    """
    Get financial health indicators.

    Args:
        start_date: Start date for analysis
        end_date: End date for analysis
        mode: Analysis mode (analysis, with_capital, or complete)
        entity_id: Entity ID to filter by
        db: Database session

    Returns:
        FinancialHealthResponse: Financial health indicators
    """
    service = AnalysisService(db)

    date_range = DateRange(start_date=start_date, end_date=end_date) if start_date or end_date else None
    result = service.get_financial_health_indicators(date_range=date_range, mode=_convert_mode(mode), entity_id=entity_id)

    return FinancialHealthResponse(
        income_to_expense_ratio=result.income_to_expense_ratio,
        savings_rate=result.savings_rate,
        average_daily_income=result.average_daily_income,
        average_daily_expense=result.average_daily_expense,
        net_daily_cash_flow=result.net_daily_cash_flow,
        period_start=result.period_start,
        period_end=result.period_end
    )

@router.get("/analysis/summary", response_model=FinancialSummaryResponse)
def get_financial_summary(
    start_date: Optional[date] = Query(None, description="Start date for analysis"),
    end_date: Optional[date] = Query(None, description="End date for analysis"),
    mode: AnalysisModeEnum = Query(AnalysisModeEnum.ANALYSIS, description="Analysis mode"),
    top_n: int = Query(5, ge=1, le=20, description="Number of top categories to return"),
    recent_count: int = Query(10, ge=1, le=50, description="Number of recent transactions to return"),
    entity_id: Optional[int] = Query(None, description="Entity ID to filter by"),
    db: Session = Depends(get_db)
):
    """
    Get comprehensive financial summary.

    This endpoint provides a complete financial overview including:
    - Total income, expenses, and net cash flow
    - Top income and expense categories
    - Recent transactions
    - Financial health indicators

    Args:
        start_date: Start date for analysis
        end_date: End date for analysis
        mode: Analysis mode (analysis or complete)
        top_n: Number of top categories to return
        recent_count: Number of recent transactions to return
        entity_id: Entity ID to filter by
        db: Database session

    Returns:
        FinancialSummaryResponse: Comprehensive financial summary
    """
    service = AnalysisService(db)

    date_range = DateRange(start_date=start_date, end_date=end_date) if start_date or end_date else None
    result = service.get_financial_summary(
        date_range=date_range,
        mode=_convert_mode(mode),
        top_n=top_n,
        recent_count=recent_count,
        entity_id=entity_id
    )

    # Convert to response model
    return FinancialSummaryResponse(
        total_income=result['total_income'],
        total_expenses=result['total_expenses'],
        net_cash_flow=result['net_cash_flow'],
        income_count=result['income_count'],
        expense_count=result['expense_count'],
        top_income_categories=[TopCategory(**cat) for cat in result['top_income_categories']],
        top_expense_categories=[TopCategory(**cat) for cat in result['top_expense_categories']],
        recent_transactions=[RecentTransaction(**txn) for txn in result['recent_transactions']],
        financial_health=FinancialHealthResponse(**result['financial_health']),
        period_start=result['period_start'],
        period_end=result['period_end'],
        mode=AnalysisModeEnum(result['mode'].value)
    )


@router.get("/analysis/income-expense", response_model=IncomeExpenseComparisonResponse)
def get_income_expense_comparison(
    start_date: Optional[date] = Query(None, description="Start date for analysis"),
    end_date: Optional[date] = Query(None, description="End date for analysis"),
    mode: AnalysisModeEnum = Query(AnalysisModeEnum.ANALYSIS, description="Analysis mode"),
    top_n: int = Query(10, ge=1, le=50, description="Number of top categories to return"),
    entity_id: Optional[int] = Query(None, description="Entity ID to filter by"),
    db: Session = Depends(get_db)
):
    """
    Get combined income and expense comparison.

    This endpoint provides a side-by-side comparison of income and expenses including:
    - Detailed income analysis
    - Detailed expense analysis
    - Cash flow summary
    - Comparison metrics (ratios, percentages)

    Args:
        start_date: Start date for analysis
        end_date: End date for analysis
        mode: Analysis mode (analysis or complete)
        top_n: Number of top categories to return
        entity_id: Entity ID to filter by
        db: Database session

    Returns:
        IncomeExpenseComparisonResponse: Income vs expense comparison
    """
    service = AnalysisService(db)

    date_range = DateRange(start_date=start_date, end_date=end_date) if start_date or end_date else None
    result = service.get_income_expense_comparison(
        date_range=date_range,
        mode=_convert_mode(mode),
        top_n=top_n,
        entity_id=entity_id
    )

    # Convert breakdown to Pydantic models for income
    income_breakdown = {
        cat: CategoryBreakdown(**data)
        for cat, data in result['income_analysis'].breakdown_by_category.items()
    }

    # Convert breakdown to Pydantic models for expenses
    expense_breakdown = {
        cat: CategoryBreakdown(**data)
        for cat, data in result['expense_analysis'].breakdown_by_category.items()
    }

    return IncomeExpenseComparisonResponse(
        income_analysis=IncomeAnalysisResponse(
            total_income=result['income_analysis'].total_income,
            transaction_count=result['income_analysis'].transaction_count,
            average_transaction=result['income_analysis'].average_transaction,
            breakdown_by_category=income_breakdown,
            period_start=result['income_analysis'].period_start,
            period_end=result['income_analysis'].period_end,
            mode=AnalysisModeEnum(result['income_analysis'].mode.value)
        ),
        expense_analysis=ExpenseAnalysisResponse(
            total_expenses=result['expense_analysis'].total_expenses,
            transaction_count=result['expense_analysis'].transaction_count,
            average_transaction=result['expense_analysis'].average_transaction,
            breakdown_by_category=expense_breakdown,
            top_categories=[TopCategory(**cat) for cat in result['expense_analysis'].top_categories],
            period_start=result['expense_analysis'].period_start,
            period_end=result['expense_analysis'].period_end,
            mode=AnalysisModeEnum(result['expense_analysis'].mode.value)
        ),
        cash_flow=CashFlowResponse(
            net_cash_flow=result['cash_flow'].net_cash_flow,
            total_income=result['cash_flow'].total_income,
            total_expenses=result['cash_flow'].total_expenses,
            income_count=result['cash_flow'].income_count,
            expense_count=result['cash_flow'].expense_count,
            period_start=result['cash_flow'].period_start,
            period_end=result['cash_flow'].period_end,
            mode=AnalysisModeEnum(result['cash_flow'].mode.value)
        ),
        comparison_metrics=result['comparison_metrics']
    )


@router.get("/analysis/category-breakdown", response_model=CategoryBreakdownResponse)
def get_category_breakdown(
    start_date: Optional[date] = Query(None, description="Start date for analysis"),
    end_date: Optional[date] = Query(None, description="End date for analysis"),
    mode: AnalysisModeEnum = Query(AnalysisModeEnum.ANALYSIS, description="Analysis mode"),
    entity_id: Optional[int] = Query(None, description="Entity ID to filter by"),
    db: Session = Depends(get_db)
):
    """
    Get detailed category breakdown for income and expenses.

    This endpoint provides a comprehensive breakdown of all categories including:
    - Income categories with totals, counts, and percentages
    - Expense categories with totals, counts, and percentages
    - Percentage of total for each category type
    - Percentage of all transactions

    Args:
        start_date: Start date for analysis
        end_date: End date for analysis
        mode: Analysis mode (analysis or complete)
        entity_id: Entity ID to filter by
        db: Database session

    Returns:
        CategoryBreakdownResponse: Detailed category breakdown
    """
    service = AnalysisService(db)

    date_range = DateRange(start_date=start_date, end_date=end_date) if start_date or end_date else None
    result = service.get_category_breakdown(
        date_range=date_range,
        mode=_convert_mode(mode),
        entity_id=entity_id
    )

    return CategoryBreakdownResponse(
        income_categories=[CategoryBreakdownItem(**cat) for cat in result['income_categories']],
        expense_categories=[CategoryBreakdownItem(**cat) for cat in result['expense_categories']],
        total_income=result['total_income'],
        total_expenses=result['total_expenses'],
        period_start=result['period_start'],
        period_end=result['period_end'],
        mode=AnalysisModeEnum(result['mode'].value)
    )


@router.get("/analysis/expenses/category-trends")
def get_expense_category_trends(
    start_date: Optional[date] = Query(None, description="Start date for analysis"),
    end_date: Optional[date] = Query(None, description="End date for analysis"),
    period: TimePeriodEnum = Query(TimePeriodEnum.MONTHLY, description="Period granularity"),
    mode: AnalysisModeEnum = Query(AnalysisModeEnum.ANALYSIS, description="Analysis mode"),
    top_n: int = Query(8, description="Number of top categories to include"),
    entity_id: Optional[int] = Query(None, description="Entity ID to filter by"),
    db: Session = Depends(get_db)
):
    """
    Get expense trends broken down by category for stacked charts.

    Returns expense amounts for each category for each time period,
    allowing for accurate stacked bar charts that show actual expenses
    per period rather than proportional distributions.
    """
    service = AnalysisService(db)
    date_range = DateRange(start_date=start_date, end_date=end_date) if start_date or end_date else None

    result = service.get_expense_category_trends(
        period=_convert_period(period),
        date_range=date_range,
        mode=_convert_mode(mode),
        top_n=top_n,
        entity_id=entity_id
    )

    return result


